src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/NativeTypes.cs

```diff
@@ -0,0 +1,193 @@
+        LDAP_OPT_SERVER_EXT_ERROR = 0x34, // Not Supported in Linux
+        LDAP_OPT_HOST_REACHABLE = 0x3E, // Not Supported in Linux
+        LDAP_OPT_PING_KEEP_ALIVE = 0x36, // Not Supported in Linux
+        LDAP_OPT_PING_WAIT_TIME = 0x37, // Not Supported in Linux
+        LDAP_OPT_PING_LIMIT = 0x38, // Not Supported in Linux
+        LDAP_OPT_DNSDOMAIN_NAME = 0x3B, // Not Supported in Linux
+        LDAP_OPT_GETDSNAME_FLAGS = 0x3D, // Not Supported in Linux
+        LDAP_OPT_PROMPT_CREDENTIALS = 0x3F, // Not Supported in Linux
+        LDAP_OPT_TCP_KEEPALIVE = 0x40, // Not Supported in Linux
+        LDAP_OPT_FAST_CONCURRENT_BIND = 0x41, // Not Supported in Linux
+        LDAP_OPT_SEND_TIMEOUT = 0x42, // Not Supported in Linux
+        LDAP_OPT_REFERRAL_CALLBACK = 0x70, // Not Supported in Linux
+        LDAP_OPT_CLIENT_CERTIFICATE = 0x80, // Not Supported in Linux
+        LDAP_OPT_SERVER_CERTIFICATE = 0x81, // Not Supported in Linux
+        LDAP_OPT_AUTO_RECONNECT = 0x91, // Not Supported in Linux
+        LDAP_OPT_SSPI_FLAGS = 0x92,
+        LDAP_OPT_SSL_INFO = 0x93, // Not Supported in Linux
+        LDAP_OPT_SIGN = 0x95,
+        LDAP_OPT_ENCRYPT = 0x96,
+        LDAP_OPT_SASL_METHOD = 0x97,
+        LDAP_OPT_AREC_EXCLUSIVE = 0x98, // Not Supported in Linux
+        LDAP_OPT_SECURITY_CONTEXT = 0x99,
+        LDAP_OPT_ROOTDSE_CACHE = 0x9a // Not Supported in Linux
+    }
+
+    internal enum ResultAll
+    {
+        LDAP_MSG_ALL = 1,
+        LDAP_MSG_RECEIVED = 2,
+        LDAP_MSG_POLLINGALL = 3 // Not Supported in Linux
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    internal sealed class LDAP_TIMEVAL
+    {
+        public int tv_sec;
+        public int tv_usec;
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    internal sealed class berval
+    {
+        public int bv_len = 0;
+        public IntPtr bv_val = IntPtr.Zero;
+
+        public berval() { }
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    internal sealed class SafeBerval
+    {
+        public int bv_len = 0;
+        public IntPtr bv_val = IntPtr.Zero;
+
+        ~SafeBerval()
+        {
+            if (bv_val != IntPtr.Zero)
+            {
+                Marshal.FreeHGlobal(bv_val);
+            }
+        }
+    }
+
+    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+    internal sealed class LdapControl
+    {
+        public IntPtr ldctl_oid = IntPtr.Zero;
+        public berval ldctl_value = null;
+        public bool ldctl_iscritical = false;
+
+        public LdapControl() { }
+    }
+
+    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+    internal struct LdapReferralCallback
+    {
+        public int sizeofcallback;
+        public QUERYFORCONNECTIONInternal query;
+        public NOTIFYOFNEWCONNECTIONInternal notify;
+        public DEREFERENCECONNECTIONInternal dereference;
+    }
+
+    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+    internal struct CRYPTOAPI_BLOB
```

Comment:
Some of these types seem Windows-specific.  CRYPTOAPI_BLOB is probably already defined somewhere under src/libraries/Common/src/Interop/Windows/... and this project should just include that same file.

Reply:
This type actually gets used by a file which is compiled for both linux and Windows: LdapConnection.cs. Of course the path for it to get called would only ever happen in Windows, given that the correspoding LdapSessionOption of that property will throw PNSE on Linux. For that reason I will move it for now on the shared file under Common.